### PR TITLE
Acquire all stream related quota and cache it locally since no more than one write can happen in parallel on stream

### PR DESCRIPTION
**Pull request by MakMukhi**
This is a small optimization upon which a follow optimization will be built to reduce lock contention.
Benchmarks: https://docs.google.com/a/google.com/spreadsheets/d/1G-p6-uTt8IHRLGXLuTR6TAxeL3I5krVf5r9C86cJPDc/edit?usp=sharing

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @MakMukhi isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
The Linux Foundation CLA GitHub bot_

**Comment by MakMukhi** [comment]
I signed it!

File: transport/http2_client.go [hunk 1]
```diff
@@ -659,44 +659,53 @@ func (t *http2Client) Write(s *Stream, hdr []byte, data []byte, opts *Options) e
...
+				// Acquire local send quota to be able to write to the controlBuf.
+				localSendQuota, err = wait(s.ctx, t.ctx, s.done, s.goAway, s.localSendQuota.acquire())
+				if err != nil {
+					return err
+				}
+			}
+			if size > streamQuota {
+				size = streamQuota
+			} // No need to do that for localSendQuota since that's only a soft limit.
+			streamQuota -= size
```
Comment (dfawley) [inline transport/http2_client.go]: I think the code would be clearer as:
```go
if tq < size {
  size = tq
}
streamQuota -= size
localSendQuota -= size
```

File: transport/transport_test.go [hunk 2]
```diff
@@ -2156,3 +2166,72 @@ func TestReadGivesSameErrorAfterAnyErrorOccurs(t *testing.T) {
...
+
+func TestPingPong64KB(t *testing.T) {
+	runPingPongTest(t, 65536)
+}
+
+func TestPingPong1MB(t *testing.T) {
+	runPingPongTest(t, 1048576)
+}
+
+func runPingPongTest(t *testing.T, msgSize int) {
```
Comment (dfawley) [inline transport/transport_test.go]: Can you add a comment to say what this test is supposed to catch that our other tests don't?

File: transport/transport_test.go [hunk 3]
```diff
@@ -2183,6 +2183,10 @@ func TestPingPong1MB(t *testing.T) {
 	runPingPongTest(t, 1048576)
 }
 
+// The continous ping-pong of data on a stream is meant to excercise the
+// flow-control logic to catch potential deadlocks.
+// These tests were added after a deadlock was caught by benchmarking
```
Comment (dfawley) [inline transport/transport_test.go]: You can leave this bit of trivia out...

Maybe say: "This is a stress-test of flow control logic"?

**Pull request merged**